### PR TITLE
Hw05

### PR DESCRIPTION
**Pull request by Vlad311010**

### CODE REVIEW

File: HW05/t1/SparseMatrixEnum.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+﻿using System.Collections;
+
```
Comment (IgorKheidorov) [inline HW05/t1/SparseMatrixEnum.cs:2]: As I see it is not used, seems that some experiments )) Please remove

File: HW05/t1/SparseMatrix.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+﻿using System.Collections;
+using MatrixIndex = (int Column, int Row);
+
+namespace t1
+{
+    public class SparseMatrix : IEnumerable<long>
+    {
+        int _columns;
+        int _rows;
+
```
Comment (IgorKheidorov) [inline HW05/t1/SparseMatrix.cs:10]: Dictionary<(int, int), long>, more clear and simple. Or at least use name MatrixIndex**es**

File: HW05/t1/SparseMatrix.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                throw new ArgumentException($"Invalid matrix dimensions. Columns: {columns}, rows: {rows}");
+
+            _rows = rows;
+            _columns = columns;
+        }
+
+        public long this[int column, int row]
+        {
+            get
+            {
```
Comment (IgorKheidorov) [inline HW05/t1/SparseMatrix.cs:25]: Check indexes for outofrange

File: HW05/t1/SparseMatrix.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                long matrixValue = 0;
+                _nonzeroElements.TryGetValue((column, row), out matrixValue);
+                return matrixValue;
+            }
+            set
```
Comment (IgorKheidorov) [inline HW05/t1/SparseMatrix.cs:30]: If you set 0 to the element which was not 0 before please remove it

File: HW05/t1/SparseMatrix.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            return GetEnumerator();
+        }
+
+        public IEnumerator<long> GetEnumerator()
+        {
+            for (int y = 0; y < _rows; y++)
+            {
+                for (int x = 0; x < _columns; x++)
+                {
+                    long value = 0;
```
Comment (IgorKheidorov) [inline HW05/t1/SparseMatrix.cs:62]: use indexer here

File: HW05/t2/ISBN13.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        public ISBN13(string isbn) 
+        {
+            if (!IsValidISBNFormat(isbn))
+                throw new ArgumentException($"Invaldi ISNB code format {isbn}");
+
+            ISBN = SimplifyISBN(isbn);    
+        }
+
+        private bool IsValidISBNFormat(string isbn)
+        {
```
Comment (IgorKheidorov) [inline HW05/t2/ISBN13.cs:18]: Move to const not to create objects each time

File: HW05/t2/ISBN13.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Text.RegularExpressions;
+
+namespace t2
+{
```
Comment (IgorKheidorov) [inline HW05/t2/ISBN13.cs:4]: Cool

File: HW05/t2/Catalog.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        public ICollection<Book> Values => _catalog.Values;
+
+        public int Count => _catalog.Count;
+
+        public bool IsReadOnly => false;
+
+        public Book this[ISBN13 isbn] 
+        {
+            get
+            {
```
Comment (IgorKheidorov) [inline HW05/t2/Catalog.cs:18]: throw an exception if no key,better use TryGetValue here and return null

File: HW05/t2/Catalog.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            // Get a set of book titles from the catalog, sorted alphabetically
+            return _catalog
+                .Select(x => x.Value.Title)
+                .OrderBy(title => title);
+        }
+
+        public IEnumerable<(string, int)> GetAuthorBookCounts()
+        {
+            return _catalog
+                .Select(KeyValue => KeyValue.Value.Authors)
```
Comment (IgorKheidorov) [inline HW05/t2/Catalog.cs:47]: Better to use GroupBy here

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Pleas fix comments

File: HW05/t1/SparseMatrixEnum.cs (hunk 1 above)
Comment (Vlad311010) [reply HW05/t1/SparseMatrixEnum.cs:2]: using statement is required for 
https://github.com/Vlad311010/homework/blob/70820679348d484ace66ed75de85e17c712398a5/HW05/t1/SparseMatrixEnum.cs#L25-L31

**Pull request merged**